arangod/Graph/Enumerators/WeightedShortestPathEnumerator.h

```diff
@@ -0,0 +1,455 @@
+   */
+  void reset(VertexRef source, VertexRef target, size_t depth = 0);
+
+  /**
+   * @brief Get the next path, if available written into the result build.
+   * The given builder will be not be cleared, this function requires a
+   * prepared builder to write into.
+   *
+   * @param result Input and output, this needs to be an open builder,
+   * where the path can be placed into.
+   * Can be empty, or an openArray, or the value of an object.
+   * Guarantee: Every returned path matches the conditions handed in via
+   * options. No path is returned twice, it is intended that paths overlap.
+   * @return true Found and written a path, result is modified.
+   * @return false No path found, result has not been changed.
+   */
+  bool getNextPath(arangodb::velocypack::Builder& result);
+
+  // The reference returned by the following call is only valid until
+  // getNextPath is called again or until the WeightedShortestPathEnumerator
+  // is destroyed or otherwise modified!
+  PathResult<ProviderType, typename ProviderType::Step> const&
+  getLastPathResult() const {
+    return _resultPath;
+  }
+
+  auto destroyEngines() -> void;
+
+  /**
+   * @brief Return statistics generated since
+   * the last time this method was called.
+   */
+  auto stealStats() -> aql::TraversalStats;
+
+  auto setForbiddenVertices(std::shared_ptr<VertexSet> forbidden)
+      -> void requires HasForbidden<PathValidatorType> {
+    _left.setForbiddenVertices(forbidden);
+    _right.setForbiddenVertices(std::move(forbidden));
+  };
+
+  auto setForbiddenEdges(std::shared_ptr<EdgeSet> forbidden)
+      -> void requires HasForbidden<PathValidatorType> {
+    _left.setForbiddenEdges(forbidden);
+    _right.setForbiddenEdges(std::move(forbidden));
+  };
+
+  auto setEmitWeight(bool flag) -> void { _emitWeight = flag; }
+
+ private:
+  [[nodiscard]] auto searchDone() const -> bool;
+  // Ensure that we have fetched all vertices in the _results list. Otherwise,
+  // we will not be able to generate the resulting path
+  auto fetchResults() -> void;
+  auto fetchResult() -> void;
+
+  // Ensure that we have more valid paths in the _result stock.
+  // May be a noop if _result is not empty.
+  auto searchMoreResults() -> void;
+
+  // Check where we want to continue our search
+  // (Left or right ball)
+  auto getBallToContinueSearch() const -> BallSearchLocation;
+
+  // In case we call this method, we know that we've already produced
+  // enough results. This flag will be checked within the "isDone" method
+  // and will provide a quick exit. Currently, this is only being used for
+  // graph searches of type "Shortest Path".
+  auto setAlgorithmFinished() -> void;
+  auto setAlgorithmUnfinished() -> void;
+  [[nodiscard]] auto isAlgorithmFinished() const -> bool;
+
+ private:
+  GraphOptions _options;
+  Ball _left;
+  Ball _right;
+
+  // Templated result list, where only valid result(s) are stored in
+  CandidatesStore _candidatesStore{};
```

Comment:
This can be simplified even further. If I understand this correctly we need at most one candidate. We always keep the smallest candidate known until the radii of both balls exceed this distance.

We can even delete the whole `CandidateStore` class.

Reply:
I see, I will have a look. Good idea!

Reply:
A further sign that splitting the two classes simplifies things!